File: system_app_back_end/areas/objects/services/inner_tasks.py
"""Checkbox lines stored in an info body — not `tasks` rows.

A line matching `☐` / `☑` (or legacy `- [ ]` / `- [x]` / `- ☐`) is an
inner task. Preferred storage is the glyph alone — no list dash. If that
info is the description target of a real task, unanimous inner state mirrors
the outer task (and the outer mark writes back onto every inner line).
"""

from __future__ import annotations

import re

from models import InformationPiece, Link, ObjectEmbed, Task, db
from areas.objects.services.object_graph import TASK_LINK_TYPE
from areas.objects.services.task_ops import ACTIVE, DONE, unmarked_status_for

_LINE = re.compile(r"^(\s*)(?:([-*])\s+)?(?:\[([ xX])\]|([☐☑]))\s?(.*)$")
_SYNCING = False


class InnerTaskLine:
    __slots__ = ("start", "end", "mark_start", "mark_end", "done", "title", "indent")

    def __init__(
        self,
        *,
        start: int,
        end: int,
        mark_start: int,
        mark_end: int,
        done: bool,
        title: str,
        indent: str,
    ) -> None:
        self.start = start
        self.end = end
        self.mark_start = mark_start
        self.mark_end = mark_end
        self.done = done
        self.title = title
        self.indent = indent


def parse_inner_task_lines(body: str) -> list[InnerTaskLine]:
    items: list[InnerTaskLine] = []
    offset = 0
    text = body or ""
    for raw in text.split("\n"):
        match = _LINE.match(raw)
        line_end = offset + len(raw)
        if match:
            indent = match.group(1) or ""
            bullet = match.group(2)
            box = match.group(3)
            glyph = match.group(4)
            title = match.group(5) or ""
            prefix_len = len(indent) + (2 if bullet is not None else 0)
            if box is not None:
                done = box.lower() == "x"
                mark_start = offset + prefix_len
                mark_end = mark_start + 3
            else:
                done = glyph == "☑"
                mark_start = offset + prefix_len
                mark_end = mark_start + 1
            items.append(
                InnerTaskLine(
                    start=offset,
                    end=line_end,
                    mark_start=mark_start,
                    mark_end=mark_end,
                    done=done,
                    title=title,
                    indent=indent,
                )
            )
        offset = line_end + 1
    return items


def inner_tasks_unanimous(body: str) -> bool | None:
    items = parse_inner_task_lines(body)
    if not items:
        return None
    if all(item.done for item in items):
        return True
    if all(not item.done for item in items):
        return False
    return None


def _render_line(item: InnerTaskLine, *, done: bool) -> str:
    mark = "☑" if done else "☐"
    title = item.title
    return f"{item.indent}{mark}{f' {title}' if title else ''}"


def set_all_inner_tasks(body: str, *, done: bool) -> str:
    items = parse_inner_task_lines(body)
    if not items:
        return body or ""
    by_start = {item.start: item for item in items}
    offset = 0
    out: list[str] = []
    text = body or ""
    for raw in text.split("\n"):
        item = by_start.get(offset)
        out.append(_render_line(item, done=done) if item else raw)
        offset += len(raw) + 1
    return "\n".join(out)


def toggle_inner_task_at(body: str, offset: int) -> str | None:
    items = parse_inner_task_lines(body)
    hit = next((item for item in items if item.mark_start <= offset < item.mark_end), None)
    if hit is None:
        hit = next((item for item in items if item.start <= offset <= item.end), None)
        if hit is None or not (hit.mark_start <= offset <= hit.mark_end):
            return None
    by_start = {item.start: item for item in items}
    cursor = 0
    out: list[str] = []
    text = body or ""
    for raw in text.split("\n"):
        item = by_start.get(cursor)
        if item is hit:
            out.append(_render_line(item, done=not item.done))
        else:
            out.append(raw)
        cursor += len(raw) + 1
    return "\n".join(out)


def info_embed_for_piece(info_id: int) -> ObjectEmbed | None:
    return ObjectEmbed.query.filter_by(type="info", information_id=int(info_id)).first()


def tasks_linked_to_info_object(object_id: int) -> list[Task]:
    rows = Link.query.filter_by(
        kind="description",
        source_type=TASK_LINK_TYPE,
        target_type="info",
        target_id=int(object_id),
    ).all()
    tasks = []
    for row in rows:
        task = db.session.get(Task, row.source_id)
        if task is None or task.archived_at is not None:
            continue
        tasks.append(task)
    return tasks


def infos_linked_from_task(task_id: int) -> list[InformationPiece]:
    rows = Link.query.filter_by(
        kind="description",
        source_type=TASK_LINK_TYPE,
        source_id=int(task_id),
        target_type="info",
    ).all()
    pieces = []
    for row in rows:
        embed = db.session.get(ObjectEmbed, row.target_id)
        if embed is None or embed.information_id is None:
            continue
        info = db.session.get(InformationPiece, embed.information_id)
        if info is None:
            continue
        pieces.append(info)
    return pieces


def sync_outer_tasks_from_info(info: InformationPiece) -> None:
    """All inner done → outer done; all inner active → outer active; mixed → skip."""
    global _SYNCING
    if _SYNCING:
        return
    embed = info_embed_for_piece(info.id)
    if embed is None:
        return
    verdict = inner_tasks_unanimous(info.body or "")
    if verdict is None:
        return
    _SYNCING = True
    try:
        for task in tasks_linked_to_info_object(embed.id):
            if task.status not in (ACTIVE, DONE):
                continue
            if verdict and task.status != DONE:
                task.status = DONE
            elif not verdict and task.status == DONE:
                task.status = unmarked_status_for(task)
    finally:
        _SYNCING = False


def sync_inner_tasks_from_outer(task: Task) -> None:
    """Outer done → every inner done; outer active → every inner active."""
    global _SYNCING
    if _SYNCING:
        return
    if not getattr(task, "id", None):
        return
    if task.status not in (ACTIVE, DONE):
        return
    done = task.status == DONE
    _SYNCING = True
    try:
        for info in infos_linked_from_task(task.id):
            next_body = set_all_inner_tasks(info.body or "", done=done)
            if next_body != (info.body or ""):
                info.body = next_body
    finally:
        _SYNCING = False

File: system_app_front_end/lib/areas/objects/data/inner_tasks.dart
/// Checkbox lines stored in an info body — not [Task] rows.
///
/// A line matching `- [ ]` / `- [x]` (or `☐` / `☑`) is an inner task.
library;

final _line = RegExp(r'^(\s*)(?:[-*]\s+\[([ xX])\]|([☐☑]))\s?(.*)$');

class InnerTaskLine {
  const InnerTaskLine({
    required this.start,
    required this.end,
    required this.markStart,
    required this.markEnd,
    required this.done,
    required this.title,
    required this.indent,
  });

  final int start;
  final int end;
  final int markStart;
  final int markEnd;
  final bool done;
  final String title;
  final String indent;
}

class InnerTaskEdit {
  const InnerTaskEdit({required this.text, required this.caret});

  final String text;
  final int caret;
}

List<InnerTaskLine> parseInnerTaskLines(String body) {
  final items = <InnerTaskLine>[];
  var offset = 0;
  final lines = body.split('\n');
  for (final raw in lines) {
    final match = _line.firstMatch(raw);
    final lineEnd = offset + raw.length;
    if (match != null) {
      final indent = match.group(1) ?? '';
      final box = match.group(2);
      final glyph = match.group(3);
      final title = match.group(4) ?? '';
      late final bool done;
      late final int markStart;
      late final int markEnd;
      if (box != null) {
        done = box.toLowerCase() == 'x';
        markStart = offset + indent.length + 2;
        markEnd = markStart + 3;
      } else {
        done = glyph == '☑';
        markStart = offset + indent.length;
        markEnd = markStart + 1;
      }
      items.add(
        InnerTaskLine(
          start: offset,
          end: lineEnd,
          markStart: markStart,
          markEnd: markEnd,
          done: done,
          title: title,
          indent: indent,
        ),
      );
    }
    offset = lineEnd + 1;
  }
  return items;
}

/// `true` all done, `false` all active, `null` none or mixed.
bool? innerTasksUnanimous(String body) {
  final items = parseInnerTaskLines(body);
  if (items.isEmpty) return null;
  if (items.every((item) => item.done)) return true;
  if (items.every((item) => !item.done)) return false;
  return null;
}

String _renderLine(InnerTaskLine item, {required bool done}) {
  final mark = done ? '[x]' : '[ ]';
  final title = item.title;
  return '${item.indent}- $mark${title.isEmpty ? '' : ' $title'}';
}

String setAllInnerTasks(String body, {required bool done}) {
  final items = parseInnerTaskLines(body);
  if (items.isEmpty) return body;
  final byStart = {for (final item in items) item.start: item};
  var offset = 0;
  final out = <String>[];
  for (final raw in body.split('\n')) {
    final item = byStart[offset];
    out.add(item == null ? raw : _renderLine(item, done: done));
    offset += raw.length + 1;
  }
  return out.join('\n');
}

InnerTaskLine? innerTaskAt(String body, int offset) {
  for (final item in parseInnerTaskLines(body)) {
    if (offset >= item.start && offset <= item.end) return item;
  }
  return null;
}

bool tapHitsInnerTaskMark(String body, int offset) {
  for (final item in parseInnerTaskLines(body)) {
    if (offset >= item.markStart && offset < item.markEnd) return true;
  }
  return false;
}

String? toggleInnerTaskAt(String body, int offset) {
  final items = parseInnerTaskLines(body);
  InnerTaskLine? hit;
  for (final item in items) {
    if (offset >= item.markStart && offset < item.markEnd) {
      hit = item;
      break;
    }
  }
  if (hit == null) return null;
  final byStart = {for (final item in items) item.start: item};
  var cursor = 0;
  final out = <String>[];
  final flipped = hit;
  for (final raw in body.split('\n')) {
    final item = byStart[cursor];
    out.add(
      item == flipped ? _renderLine(flipped, done: !flipped.done) : raw,
    );
    cursor += raw.length + 1;
  }
  return out.join('\n');
}

/// Combined title\\nbody helpers — inner tasks live in the body only.

(String title, String body) _split(String combined) {
  final nl = combined.indexOf('\n');
  if (nl < 0) return (combined, '');
  return (combined.substring(0, nl), combined.substring(nl + 1));
}

String _join(String title, String body) {
  if (body.isEmpty) return title;
  return '$title\n$body';
}

int _bodyOffset(String combined) {
  final nl = combined.indexOf('\n');
  return nl < 0 ? -1 : nl + 1;
}

bool tapHitsCombinedInnerMark(String combined, int offset) {
  final bodyAt = _bodyOffset(combined);
  if (bodyAt < 0 || offset < bodyAt) return false;
  return tapHitsInnerTaskMark(combined.substring(bodyAt), offset - bodyAt);
}

String? toggleCombinedInnerTask(String combined, int offset) {
  final bodyAt = _bodyOffset(combined);
  if (bodyAt < 0 || offset < bodyAt) return null;
  final next = toggleInnerTaskAt(combined.substring(bodyAt), offset - bodyAt);
  if (next == null) return null;
  return _join(combined.substring(0, bodyAt - 1), next);
}

String setAllCombinedInnerTasks(String combined, {required bool done}) {
  final parts = _split(combined);
  return _join(parts.$1, setAllInnerTasks(parts.$2, done: done));
}

bool? combinedInnerTasksUnanimous(String combined) {
  return innerTasksUnanimous(_split(combined).$2);
}

/// `- ` / `* ` at the start of a body line becomes `- [ ] `.
InnerTaskEdit? promoteBareDashToCheckbox(String combined, int caret) {
  final bodyAt = _bodyOffset(combined);
  if (bodyAt < 0 || caret < bodyAt) return null;
  final body = combined.substring(bodyAt);
  final local = caret - bodyAt;
  final lineStart = body.lastIndexOf('\n', local - 1) + 1;
  final lineEnd = body.indexOf('\n', lineStart);
  final end = lineEnd < 0 ? body.length : lineEnd;
  final line = body.substring(lineStart, end);
  final match = RegExp(r'^(\s*)([-*])\s$').firstMatch(line);
  if (match == null) return null;
  if (local != lineStart + line.length) return null;
  final indent = match.group(1) ?? '';
  final nextLine = '$indent- [ ] ';
  final nextBody = body.replaceRange(lineStart, end, nextLine);
  return InnerTaskEdit(
    text: _join(combined.substring(0, bodyAt - 1), nextBody),
    caret: bodyAt + lineStart + nextLine.length,
  );
}

/// Plain Enter on a checkbox line: next item, or drop an empty prefix.
InnerTaskEdit? insertInnerTaskLineOnEnter(String combined, int caret) {
  final bodyAt = _bodyOffset(combined);
  if (bodyAt < 0 || caret < bodyAt) return null;
  final body = combined.substring(bodyAt);
  final local = (caret - bodyAt).clamp(0, body.length);
  final item = innerTaskAt(body, local);
  if (item == null) return null;
  if (item.title.trim().isEmpty) {
    final title = combined.substring(0, bodyAt - 1);
    final nextBody = body.replaceRange(item.start, item.end, item.indent);
    if (nextBody.isEmpty) {
      return InnerTaskEdit(text: '$title\n', caret: title.length + 1);
    }
    return InnerTaskEdit(
      text: _join(title, nextBody),
      caret: bodyAt + item.start + item.indent.length,
    );
  }
  final insertAt = item.end;
  const fresh = '- [ ] ';
  final prefix = insertAt < body.length ? '\n$fresh' : '\n$fresh';
  final nextBody = body.replaceRange(insertAt, insertAt, prefix);
  return InnerTaskEdit(
    text: _join(combined.substring(0, bodyAt - 1), nextBody),
    caret: bodyAt + insertAt + prefix.length,
  );
}

/// Backspace on an empty checkbox (or at the mark) removes the prefix.
InnerTaskEdit? backspaceInnerTaskPrefix(String combined, int caret) {
  final bodyAt = _bodyOffset(combined);
  if (bodyAt < 0 || caret < bodyAt) return null;
  final body = combined.substring(bodyAt);
  final local = caret - bodyAt;
  final item = innerTaskAt(body, local);
  if (item == null) return null;
  final atMark = local <= item.markEnd && local >= item.start;
  if (!atMark && !(item.title.isEmpty && local == item.end)) return null;
  if (item.title.isNotEmpty && local > item.markEnd) return null;
  final title = combined.substring(0, bodyAt - 1);
  final nextBody = body.replaceRange(item.start, item.end, item.indent);
  if (nextBody.isEmpty) {
    return InnerTaskEdit(text: '$title\n', caret: title.length + 1);
  }
  return InnerTaskEdit(
    text: _join(title, nextBody),
    caret: bodyAt + item.start + item.indent.length,
  );
}
